### PR TITLE
refactor(rust): apply cli_state abstraction

### LINKED ISSUE #4410: Create traits to abstract the configuration files handled by the CLI state
### Current behavior

Currently the different resources within the CLI State are all implemented independently but share a very similar structure.

Those CLI State resources are used and defined within [here](https://github.com/build-trust/ockam/blob/1ef02c9c07603b5de776563e6861cfdf1aae3dce/implementations/rust/ockam/ockam_api/src/cli_state.rs#L60-L67).

### Desired behavior

Abstract many of the existing systems and create a common structure for the way CLI State resources are persisted and accessed.

### PR DESCRIPTION
**Pull request by adrianbenavides**
Fixes https://github.com/build-trust/ockam/issues/4410

### CODE REVIEW

File: implementations/rust/ockam/ockam_api/src/cli_state/identities.rs
Comment (etorreborre) [inline implementations/rust/ockam/ockam_api/src/cli_state/identities.rs]: Note: those 2 fields could be replaced with `pub identity: Identity`

File: implementations/rust/ockam/ockam_api/src/cli_state/traits.rs
Comment (etorreborre) [inline implementations/rust/ockam/ockam_api/src/cli_state/traits.rs]: It would nice to extend the description a bit to give a better mental model of what's going on:

 - the `StateDirTrait` represents a list of items, uniquely identified by a name and represented by the same `Item` type
 - all the items are persisted under the same directory
 - one item is marked as being the "default" item. It is either
   - the first created item
   - one of the existing items, marked as being the default one with the `set_default` method

Additionally some unit test would show what happens when you create/list/set_default/overwrite etc...

File: implementations/rust/ockam/ockam_api/src/cli_state/traits.rs
Comment (etorreborre) [inline implementations/rust/ockam/ockam_api/src/cli_state/traits.rs]: Just `s.delete()` without the final `Ok(())` should work here.

File: implementations/rust/ockam/ockam_api/src/cli_state/traits.rs
Comment (etorreborre) [inline implementations/rust/ockam/ockam_api/src/cli_state/traits.rs]: I am not sure if this method is absolutely needed. The existence of a default should be guaranteed by the fact that at least one item has been created. So `is_empty` would be a more direct way to test for this condition IMO.

File: implementations/rust/ockam/ockam_api/src/cli_state/traits.rs
Comment (etorreborre) [inline implementations/rust/ockam/ockam_api/src/cli_state/traits.rs]: Can we move this function to `VaultsState` if it is only used there for now?

File: implementations/rust/ockam/ockam_api/src/cli_state/traits.rs
Comment (etorreborre) [inline implementations/rust/ockam/ockam_api/src/cli_state/traits.rs:170]: This is fishy. An item shouldn't have to be able to access the global state. If we have a look at where it used, it is used to check a dependency between identity and node when trying to delete an identity. 

I propose that:

 - the code for deleting an identity state does just that, and does not try to check if a running node uses it with `identity.in_use` (which btw throws an error while the name implies that it should just return a boolean)
 
 - the `DeleteCommand` which has access to the global `CliState` calls a function on that state to check the dependency between the identity state and the nodes state, then decides if it's okay to call `identity_state.delete()`

This makes the surface of the the `StateItemTrait` smaller and makes each state code more independent of other state code.

File: implementations/rust/ockam/ockam_api/src/cli_state/traits.rs
Comment (etorreborre) [inline implementations/rust/ockam/ockam_api/src/cli_state/traits.rs:196]: We shouldn't have to expose the `path` and `data_path` functions. The purpose of the `StateItem` trait is to manage the persistence of data somewhere and return some well-formed data as `Config`. 

For example in the case of a `VaultState` I think that `Config` should be a `VaultConfig` to be fed directly into a `Vault::create_from_config(vault_config)` function and the implementation of `VaultState` should sort out that the persisted data needs to a be a json file + some storage file. By exposing `path` and `data_path` we are breaking the encapsulation that we are looking for with those traits because a user of that path needs to know exactly what they can or cannot do with it.

File: implementations/rust/ockam/ockam_api/src/cli_state/traits.rs
Comment (etorreborre) [inline implementations/rust/ockam/ockam_api/src/cli_state/traits.rs:184]: We could get rid of the `load` method by caching the `Config` on the first access to the `config` method.

File: implementations/rust/ockam/ockam_api/src/cli_state/traits.rs
Comment (etorreborre) [inline implementations/rust/ockam/ockam_api/src/cli_state/traits.rs]: This function is not used.

File: implementations/rust/ockam/ockam_api/src/cli_state/traits.rs
Comment (etorreborre) [inline implementations/rust/ockam/ockam_api/src/cli_state/traits.rs]: I think we should make a clear distinction between:

 - a template / some helper functions to load / persist a `StateDir` and not expose the various `dir` and `path` functions. If there is a `data` dir for a given implementation, the `load` function can do the `std::fs::create_dir_all(dir.join("data"))?;` by itself and maybe use `DATA: &str = "data"` as a constant so we make sure that we are consistent in terms of naming
  
 - the `StateDir` interface which is basically
   - `create(name, config)`
   - `overwrite(name, config)`  
   - `list()`
   - `get(name)`
   - `delete(name)`

**Review by etorreborre — COMMENTED** [review]
I think that this is step in the right direction but I also think that we should absolutely go one step further:

 - keep all these paths hidden as implementation details
 - use well-typed interfaces to `list, get, delete,...` well-typed configs 

@adrianbenavides what I wrote above still implies a lot of work so feel free to leave it to further PRs

**Comment by adrianbenavides** [comment]
Thanks @etorreborre for the review, that was very helpful 🙏 ! I'll apply the most superficial changes, and I'll tackle the rest in a separate PR so we can also unblock the PRs that depend on this one.

**Comment by adrianbenavides** [comment]
I've addressed a bunch of the comments (closed now) and created [an issue](https://github.com/build-trust/ockam/issues/4829) with all the comments I didn't fix. Feel free to close the comments if you think that's okey.

**Review by etorreborre — APPROVED** [review]
Great, let's tackle improvements in some follow-up PRs

**Pull request merged**
